File: hochbaumShmyos.py
import numpy as np
import pandas as pd

def hochbaum_shmoys_k_center_from_csv(df, k):
    # Extrahiere die Koordinatenpunkte aus dem DataFrame
    points = df[['x', 'y']].values
    
    # Wähle den ersten Punkt zufällig aus den Eingabepunkten
    centers = [points[np.random.randint(len(points))]]
    
    # Berechne den maximalen Abstand für die initiale Zuordnung
    distances = np.array([np.linalg.norm(point - centers[0]) for point in points])
    max_distance = np.max(distances)

    # Iteriere, um die restlichen k - 1 Zentren zu finden
    while len(centers) < k:
        # Wähle das nächste Zentrum basierend auf dem größten Abstand
        next_center_index = np.argmax(distances)
        next_center = points[next_center_index]
        centers.append(next_center)
        
        # Aktualisiere die Abstände für jeden Punkt zu seinem nächsten Zentrum
        for i, point in enumerate(points):
            distances[i] = min(distances[i], np.linalg.norm(point - next_center))
        
        # Berechne den neuen maximalen Abstand
        max_distance = np.max(distances)
    
    return max_distance

File: hochbaumShmyos.cpp
#include "header/hochbaumShmyos.h"
#include "header/Point.h"
#include <algorithm>

using namespace std;

double hochbaumShmoysKCenter(const vector<Point> &points, int k) {
  vector<Point> centers;

  // Wähle den ersten Punkt zufällig aus den Eingabepunkten
  centers.push_back(points[rand() % points.size()]);

  vector<double> distances(points.size());

  // Berechne den maximalen Abstand für die initiale Zuordnung
  double maxDistance = 0.0;
  for (int i = 0; i < points.size(); ++i) {
    distances[i] = Point::distance(points[i], centers[0]);
    maxDistance = max(maxDistance, distances[i]);
  }

  // Iteriere, um die restlichen k - 1 Zentren zu finden
  while (centers.size() < k) {
    // Wähle das nächste Zentrum basierend auf dem größten Abstand
    int nextCenterIndex = distance(distances.begin(), max_element(distances.begin(), distances.end()));
    Point nextCenter = points[nextCenterIndex];
    centers.push_back(nextCenter);

    // Aktualisiere die Abstände für jeden Punkt zu seinem nächsten Zentrum
    // Berechne den neuen maximalen Abstand
    for (int i = 0; i < points.size(); ++i) {
      distances[i] = min(distances[i], Point::distance(points[i], nextCenter));
      maxDistance = max(maxDistance, distances[i]);
    }
  }

  return maxDistance;
}